homeassistant/components/trafikverket_train/coordinator.py

```diff
@@ -0,0 +1,128 @@
+
+
+class TVDataUpdateCoordinator(DataUpdateCoordinator):
+    """A Trafikverket Data Update Coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        to_station: StationInfo,
+        from_station: StationInfo,
+    ) -> None:
+        """Initialize the Trafikverket coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=TIME_BETWEEN_UPDATES,
+        )
+        self._train_api = TrafikverketTrain(
+            async_get_clientsession(hass), entry.data[CONF_API_KEY]
+        )
+        self._from: StationInfo = from_station
+        self._to: StationInfo = to_station
+        self._time: time | None = dt_util.parse_time(entry.data[CONF_TIME])
+        self._weekdays: list[str] = entry.data[CONF_WEEKDAY]
+
+    async def _async_update_data(self) -> dict[str, StateType | datetime | timedelta]:
+        """Fetch data from Trafikverket."""
+
+        when = dt_util.now()
+        state: TrainStop | None = None
+        if self._time:
+            departure_day = next_departuredate(self._weekdays)
+            when = datetime.combine(
+                departure_day,
+                self._time,
+                dt_util.get_time_zone(self.hass.config.time_zone),
+            )
+        try:
+            if self._time:
+                state = await self._train_api.async_get_train_stop(
+                    self._from, self._to, when
+                )
+            else:
+                state = await self._train_api.async_get_next_train_stop(
+                    self._from, self._to, when
+                )
+        except ValueError as error:
+            if "Invalid authentication" in error.args[0]:
+                raise ConfigEntryAuthFailed from error
+            raise UpdateFailed(
+                f"Train departure {when} encountered a problem: {error}"
+            ) from error
+
+        departure_time = state.advertised_time_at_location
+        if state.time_at_location:
+            departure_time = state.time_at_location
+        if state.estimated_time_at_location:
+            departure_time = state.estimated_time_at_location
+
+        states: dict[str, StateType | datetime | timedelta] = {
+            "departure_time": dt_util.as_utc(departure_time)
+            if departure_time
+            else None,
+            "departure_state": state.get_state().value,
+            "cancelled": state.canceled,
+            "delayed_time": state.get_delay_time(),
+            "planned_time": dt_util.as_utc(state.advertised_time_at_location)
+            if state.advertised_time_at_location
+            else None,
+            "estimated_time": dt_util.as_utc(state.estimated_time_at_location)
+            if state.estimated_time_at_location
+            else None,
+            "actual_time": dt_util.as_utc(state.time_at_location)
+            if state.time_at_location
+            else None,
```

Comment:
Would it maybe be more readable if you extract the `.as_utc` and `", ".join` into small separate helper functions? It would clean up this dict a lot imo